src/communication/v2/pool.hpp

```diff
@@ -69,23 +59,16 @@ class IOContextPool {
     background_threads_.clear();
   }
 
-  boost::asio::io_context &GetIOContext() {
-    // Use a round-robin scheme to choose the next io_service to use.
-    auto &io_context = io_contexts_[next_io_context_++];
-    if (next_io_context_ == io_contexts_.size()) {
-      next_io_context_ = 0;
-    }
-    return *io_context;
-  }
-
   bool IsRunning() const noexcept { return running_; }
 
+  IOContext &GetIOContext() noexcept { return io_context_; }
+
  private:
   /// The pool of io_context.
-  std::vector<std::shared_ptr<IOContext>> io_contexts_;
-  std::vector<std::shared_ptr<IOContextGuard>> guards_;
+  IOContext io_context_;
+  IOContextGuard guard_;
+  size_t pool_size_;
   std::vector<std::thread> background_threads_;
```

Comment:
Maybe `std::jthread` can be used, so the joining would be done in a RAII manner, `AwaitShutdown` should become much simpler, just `background_threads_.clear()`.

Reply:
This is a gem :gem: